sources/packages/backend/apps/api/src/route-controllers/announcements/_tests_/e2e/announcement.institutions.controller.getAnnouncements.e2e-spec.ts

```diff
@@ -0,0 +1,127 @@
+import { HttpStatus, INestApplication } from "@nestjs/common";
+import * as request from "supertest";
+import { Repository } from "typeorm";
+import { Announcement } from "@sims/sims-db";
+import { addDays } from "@sims/utilities";
+import {
+  BEARER_AUTH_TYPE,
+  createTestingAppModule,
+  getInstitutionToken,
+  InstitutionTokenTypes,
+} from "../../../../testHelpers";
+
+describe("AnnouncementInstitutionsController(e2e)-getAnnouncements", () => {
+  let app: INestApplication;
+  let announcementsRepo: Repository<Announcement>;
+  const endpoint = `/institutions/announcements?target=institution-dashboard`;
+
+  beforeAll(async () => {
+    const { nestApplication, dataSource } = await createTestingAppModule();
+    app = nestApplication;
+    announcementsRepo = dataSource.getRepository(Announcement);
+  });
+
+  it("Should return a current announcement.", async () => {
+    // Arrange
+    const now = new Date();
+    const before = addDays(-10, now);
+    const future = addDays(30, now);
+
+    const announcement = new Announcement();
+    announcement.message = "test announcement";
+    announcement.messageTitle = "test title";
+    announcement.target = ["student-dashboard", "institution-dashboard"];
+    announcement.startDate = before;
+    announcement.endDate = future;
+    await announcementsRepo.save(announcement);
+    const institutionUserToken = await getInstitutionToken(
+      InstitutionTokenTypes.CollegeFUser,
+    );
+
+    // Act/Assert
+    const response = await request(app.getHttpServer())
+      .get(endpoint)
+      .auth(institutionUserToken, BEARER_AUTH_TYPE)
+      .expect(HttpStatus.OK);
+
+    expect(
+      response.body.some(
+        (announcement) =>
+          announcement.message === "test announcement" &&
+          announcement.messageTitle === "test title" &&
+          announcement.startDate === before.toISOString() &&
+          announcement.endDate === future.toISOString(),
+      ),
+    ).toBe(true);
+  });
+
+  it("Should not return an outdated announcement.", async () => {
+    // Arrange
+    const now = new Date();
+    const before = addDays(-20, now);
+
+    const announcement = new Announcement();
+    announcement.message = "already done";
+    announcement.messageTitle = "already done";
+    announcement.target = ["student-dashboard", "institution-dashboard"];
+    announcement.startDate = before;
+    announcement.endDate = before;
+    await announcementsRepo.save(announcement);
+    const institutionUserToken = await getInstitutionToken(
+      InstitutionTokenTypes.CollegeFUser,
+    );
+
+    // Act/Assert
+    const response = await request(app.getHttpServer())
+      .get(endpoint)
+      .auth(institutionUserToken, BEARER_AUTH_TYPE)
+      .expect(HttpStatus.OK);
+
+    expect(
+      response.body.some(
```

Comment:
The assert could be simplified and used like this. await request(app.getHttpServer())
      .get(endpoint)
      .auth(institutionUserToken, BEARER_AUTH_TYPE)
      .expect(HttpStatus.OK)
      .expect([
        {
          messageTitle: announcement.messageTitle,
          message: announcement.message,
          startDate: announcement.startDate.toISOString(),
          endDate: announcement.endDate.toISOString(),
          target: announcement.target,
        },
      ]);

Reply:
There's nothing saying that that will be the only announcement returned, but I have updated the test to refer to the faked announcement's values directly.

Reply:
Generally when writing tests, we make sure that API or any portion of the test returns only the data that is created for that test. 

e.g. if we create 1 single announcement which is valid for current date in any test, we must expect that announcement to be returned, so that we do not have to check if that announcement is one one among many valid ones.

Reply:
Unless I am allowed to delete / modify other announcements that may already be in the db, I can not guarantee that there will not already be relevant announcements in the db. The get for announcements does not fetch _this specific announcement_, it fetches anything in the correct time range and dashboard.

Reply:
Whatever you mentioned above is absolutely right, but only if you repeat the target. 

If you use a distinct target for each test case, then the data what we expect can be guaranteed. 
For e.g. for the first test case, I can create a target called `fake-institution-target-1` and expect a guaranteed result like this.  Please keep in mind that, we are testing the API and target(s) doesn't need to be what we use in UI.

```ts

   const now = new Date();
    const before = addDays(-10, now);
    const future = addDays(30, now);

    const announcement = createFakeAnnouncement({
      initialValues: {
        startDate: before,
        endDate: future,
        target: ["fake-institution-target-1"],
      },
    });
    const endpoint = `/institutions/announcements?target=fake-institution-target-1`;
    await announcementsRepo.save(announcement);
    const institutionUserToken = await getInstitutionToken(
      InstitutionTokenTypes.CollegeFUser,
    );

    // Act/Assert
    await request(app.getHttpServer())
      .get(endpoint)
      .auth(institutionUserToken, BEARER_AUTH_TYPE)
      .expect(HttpStatus.OK)
      .expect([
        {
          messageTitle: announcement.messageTitle,
          message: announcement.message,
          startDate: announcement.startDate.toISOString(),
          endDate: announcement.endDate.toISOString(),
          target: announcement.target,
        },
      ]);

```

Reply:
I have specific targets bound to specific permissions - wouldn't this also require that to change?

`
export class StudentAnnouncementsAPIInDTO {
  @IsIn(["student-dashboard"])
  target: string;
}

export class InstitutionAnnouncementsAPIInDTO {
  @IsIn(["institution-dashboard"])
  target: string;
}
`